File: nodejs/mcma-worker/lib/builders/worker-builder.js
const OperationHandlerBuilder = require("./operation-handler-builder");
const Worker = require("../worker");

class WorkerBuilder {
    constructor() {
        const operationHandlerBuilders = [];
        
        this.handleOperation = (operationName, configureOperation) => {
            if (!configureOperation) {
                if (typeof operationName === "function" && operationName.name !== "Function") {
                    configureOperation = x => x.handle(operationName);
                    operationName = operationName.name;
                }
            }
            if (!operationName || typeof operationName !== "string" || operationName.length === "") {
                throw new Error("operationName must be a non-empty string.");
            }
            if (!configureOperation || typeof configureOperation !== "function") {
                throw new Error("configureOperation must be a function");
            }
            const opHandlerBuilder = new OperationHandlerBuilder(operationName);
            operationHandlerBuilders.push(opHandlerBuilder);
            configureOperation(opHandlerBuilder);
            return this;
        };

        this.build = () => new Worker(operationHandlerBuilders.reduce((flattened, opBuilder) => flattened.concat(opBuilder.build()), []));
    }
}

module.exports = {
    WorkerBuilder
};

File: dotnet/base/Mcma.Worker/Builders/WorkerBuilder.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Mcma.Worker.Builders
{
    public class WorkerBuilder
    {
        private List<IOperationHandlerBuilder> OperationHandlerBuilders { get; } = new List<IOperationHandlerBuilder>();

        public WorkerBuilder HandleOperation<THandler>(string operationName = null) where THandler : IWorkerOperationHandler, new()
            => HandleOperation(new THandler(), operationName);

        public WorkerBuilder HandleOperation<THandler>(THandler handler, string operationName = null) where THandler : IWorkerOperationHandler
            => HandleOperation(handler.InputType, operationName ?? typeof(THandler).Name, x => x.Handle(handler));

        public WorkerBuilder HandleOperation(Type inputType, string operationName, Action<OperationHandlerBuilder> configureOperation)
        {
            var opHandlerBuilder = new OperationHandlerBuilder(inputType, operationName);
            OperationHandlerBuilders.Add(opHandlerBuilder);
            configureOperation(opHandlerBuilder);
            return this;
        }

        public WorkerBuilder HandleOperation<T>(string operationName, Action<OperationHandlerBuilder<T>> configureOperation)
        {
            var opHandlerBuilder = new OperationHandlerBuilder<T>(operationName);
            OperationHandlerBuilders.Add(opHandlerBuilder);
            configureOperation(opHandlerBuilder);
            return this;
        }

        public IWorker Build() => new Worker(OperationHandlerBuilders.SelectMany(r => r.Build()));
    }
}
